cou/steps/analyze.py

```diff
@@ -0,0 +1,205 @@
+# Copyright 2023 Canonical Limited.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Functions for analyzing an OpenStack cloud before an upgrade."""
+from __future__ import annotations
+
+import logging
+from collections import defaultdict
+from dataclasses import dataclass, field
+from typing import Any, Dict, Iterable, Union
+
+import yaml
+from juju.client._definitions import ApplicationStatus
+
+from cou.exceptions import CommandRunFailed
+from cou.utils.juju_utils import (
+    async_get_application_config,
+    async_get_full_juju_status,
+    async_run_on_unit,
+)
+from cou.utils.openstack import CHARM_TYPES, get_os_code_info
+from cou.utils.upgrade_utils import extract_charm_name_from_url
+
+
+@dataclass
+class Analysis:
+    """Analyze result."""
+
+    apps: Iterable[Application]
+
+    def dump(self) -> str:
+        """Dump as yaml."""
+        return "\n".join([str(app) for app in self.apps])
+
+
+@dataclass
+class Application:
+    """Representation of an application in the deployment."""
+
+    # pylint: disable=too-many-instance-attributes
+
+    name: str
+    status: ApplicationStatus
+    config: dict
+    model_name: str
+    charm: str = ""
+    charm_origin: str = ""
+    os_origin: str = ""
+    channel: str = ""
+    pkg_name: str = ""
+    # E.g of units: {"keystone/0": {'os_version': 'victoria', 'pkg_version': '2:18.1'}}
+    units: defaultdict[str, dict] = field(default_factory=lambda: defaultdict(dict))
+    #  E.g of os_release_units: {"ussuri":{"keystone/0"}, "victoria": {"keystone/1"}}
+    os_release_units: defaultdict[str, set] = field(default_factory=lambda: defaultdict(set))
+    #  E.g of pkg_version_units: {"2:17.0": {"keystone/0"}, "2:18.0": {"keystone/1"}}
+    pkg_version_units: defaultdict[str, set] = field(default_factory=lambda: defaultdict(set))
```

Comment:
Looking again at those information, it's not necessary to have `pkg_version_units` and I'll delete then from the code. However, `os_release_units` it's useful to check if an application has more than one OpenStack version by checking the len of `os_release_units` (this will be used on the planning part). We could access this same information by looping units, but IMO it's easier to access this way.
Regarding `cache coherence issues` I think it's very unlikely to happen. Both `units` and `os_release_units` are filled in the same [for loop](https://github.com/gabrielcocenza/charmed-openstack-upgrader-1/blob/remove-zaza-utils-final/cou/steps/analyze.py#L111) that use the same status when the object is instantiate.

Reply:
sure, but I think it would be easier to start from a minimal data collection and then potentially expand it, rather than collect anything we think may be useful and then remove things we don't need. Until we hit a performance bottleneck (and I doubt we ever will), I think it'd be safer to stick to one or the other variable.
For example, you could start with only `os_release_units` (and rename it to something simpler); I suspect that's going to be the more useful of the two.